design/resource-management.md

```diff
@@ -0,0 +1,329 @@
+constrained, measured and not accounted for, the node will run into CPU and memory pressure unexpectedly.
+
+#### Consistent guaranteed pod behavior
+
+Predictable performance is important for end users. By pushing IO threads into a shared pool, the
+achievable performance will be inconsistent.  Even if a user utilizes a `guaranteed` QoS pod, the
+performance profile will differ depending on the amount of contention on the system.  Raw unconstrained
+performance is important for Kata, but not as important as consistent and predictable behavior.
+
+#### OOM, unbound CPU utilization
+
+Memory limits are enforced, not requests. Until [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md)
+is added, users or admission controllers need to provide higher limits.
+
+In the case of CRI-O, the memory is being charged to conmon which is bounded by pod limits. As a result,
+the workload can still be OOMed. This is okay from a node stability point of view, but not from a pod stability
+point of view. This behavior assumes we are called by conmon and that they are constrained appropriately. Luckily,
+this is reasonably correct from a memory point of view. I/O bound workloads will exhibit sub-optimal performance
+due to the CPU constraints applied to conmon (where the io threads run).
+
+For containerd, the memory is being charged to containerd, which is basically unbounded.  This is bad for
+node stability, as the `pod` is essentially unbounded.
+
+## Proposed Changes
+
+### Summary
+ * Pause cgroup cpu shaes should be setup correctly.
+ * Do not create container cgroups on the host. Instead, create a pod sandbox group that is entirely managed by Kata
+ * Move the QEMU threads into the sandbox cgroup
+
+With these changes, performance and constraints for a pod is consistent. This constraining change will be
+more restrictive relative to existing design.
+
+The overheads associated with running a sandbox should be accounted for explicitly, and at the pod level.
+Once the Pod Overhead KEP is available, this should become a part of RuntimeClass, applied to pods which
+utilize the applicable RuntimeClass.  See [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md)
+for more details.
+
+### Details
+
+#### Pod Sandbox Cgroup
+ * The pod sandbox cgroup should always be the summation of all container group resources.
+ * In the case of cri-o where it creates other cgroups for conmon, they will be siblings.
+ * The conmon cgroups today are rather large, so if conmon goes wild there is a possibility
+ the workload will get fewer resources. But it will not introduce any other side effects.
+
+## Alternatives Considered
+
+### Only constrain vCPUs, leaving remaining threads for system reserved
+
+This will, in some cases, provide improved performance.
```

Comment:
The problem with system reserved is that it does not scale. So I don't think we should put scaling factors (qemu main thread, vhost etc.) there and it can result in unexpected system failures due to enforced constraints in system reserved.